quill-core-portable/src/main/scala/io/getquill/quat/Quat.scala

```diff
@@ -0,0 +1,326 @@
+package io.getquill.quat
+
+import io.getquill.quotation.QuatException
+
+import scala.collection.mutable
+
+object LinkedHashMapOps {
+  implicit class LinkedHashMapExt[K, V](m1: mutable.LinkedHashMap[K, V]) {
+    def zipWith[R](m2: mutable.LinkedHashMap[K, V])(f: PartialFunction[(K, V, Option[V]), R]) =
+      LinkedHashMapOps.zipWith(m1, m2, f)
+  }
+
+  def zipWith[K, V, R](m1: mutable.LinkedHashMap[K, V], m2: mutable.LinkedHashMap[K, V], f: PartialFunction[(K, V, Option[V]), R]) = {
+    m1.iterator.map(r => (r._1, r._2, m2.get(r._1))).collect(f)
+  }
+}
+
+// This represents a simplified Sql-Type. Since it applies to all dialects, it is called
+// Quill-Application-Type hence Quat.
+sealed trait Quat {
+  def applyRenames: Quat = this
+  def withRenames(renames: List[(String, String)]): Quat
+
+  def renames: List[(String, String)] = List()
+
+  /** Either convert to a Product or make the Quat into an error if it is anything else. */
+  def probit =
+    this match {
+      case p: Quat.Product => p
+      case e: Quat.Error   => e
+      case other           => Quat.Error(s"Was expecting SQL-level type must be a Product but found `${other}`")
+    }
+
+  def leastUpperType(other: Quat): Option[Quat] = {
+    (this, other) match {
+      case (Quat.Generic, other)                   => Some(other)
+      case (Quat.Null, other)                      => Some(other)
+      case (other, Quat.Generic)                   => Some(other)
+      case (other, Quat.Null)                      => Some(other)
+      case (Quat.Value, Quat.Value)                => Some(Quat.Value)
+      case (me: Quat.Product, other: Quat.Product) => me.leastUpperTypeProduct(other)
+      case (_, _)                                  => None
+    }
+  }
+
+  def leastUpperType2(other: Quat): Option[Quat] = {
```

Comment:
Is this still necessary? I think it can be ✂️ ?

Reply:
That was actually a stand-in for a future feature. I am using this now and not the other one.